llarp/router/outbound_message_handler.cpp

```diff
@@ -0,0 +1,219 @@
+#include <router/outbound_message_handler.hpp>
+
+#include <messages/link_message.hpp>
+#include <router/i_outbound_session_maker.hpp>
+#include <link/i_link_manager.hpp>
+#include <constants/link_layer.hpp>
+#include <util/memfn.hpp>
+#include <util/status.hpp>
+
+#include <algorithm>
+#include <cstdlib>
+
+namespace llarp
+{
+  bool
+  OutboundMessageHandler::QueueMessage(const RouterID &remote,
+                                       const ILinkMessage *msg,
+                                       SendStatusHandler callback)
+  {
+    std::array< byte_t, MAX_LINK_MSG_SIZE > linkmsg_buffer;
+    llarp_buffer_t buf(linkmsg_buffer);
+
+    if(!EncodeBuffer(msg, buf))
+    {
+      return false;
+    }
+
+    buf.sz = buf.cur - buf.base;
+
+    Message message{buf.sz, callback};
+
+    std::copy_n(buf.base, buf.sz, message.first.data());
+
+    if(SendIfSession(remote, message))
+    {
+      return true;
+    }
+
+    bool shouldCreateSession = false;
+    {
+      util::Lock l(&_mutex);
+
+      // create queue for <remote> if it doesn't exist, and get iterator
+      auto itr_pair = outboundMessageQueue.emplace(remote, MessageQueue());
+
+      itr_pair.first->second.push(std::move(message));
+
+      shouldCreateSession = itr_pair.second;
+    }
+
+    if(shouldCreateSession)
+    {
+      QueueSessionCreation(remote);
+    }
+
+    return true;
+  }
+
+  // TODO: this
+  util::StatusObject
+  OutboundMessageHandler::ExtractStatus() const
+  {
+    util::StatusObject status{};
+    return status;
+  }
+
+  void
+  OutboundMessageHandler::Init(ILinkManager *linkManager,
+                               std::shared_ptr< Logic > logic)
+  {
+    _linkManager = linkManager;
+    _logic       = logic;
+  }
+
+  void
+  OutboundMessageHandler::OnSessionEstablished(const RouterID &router)
+  {
+    FinalizeRequest(router, SendStatus::Success);
+  }
+
+  void
+  OutboundMessageHandler::OnConnectFailure(const RouterID &router)
+  {
+    FinalizeRequest(router, SendStatus::Timeout);
```

Comment:
is a connect failure always a timeout?

Reply:
when this func is called, yes.  Changing func name.